messaging/src/main/java/org/axonframework/queryhandling/ReactorQueryGateway.java

```diff
@@ -0,0 +1,547 @@
+     * from a single source. Execution may be asynchronous, depending on the {@code QueryBus} implementation.
+     * <p><b>Do note that the {@code query} will not be dispatched until there is a subscription to the resulting {@link
+     * Mono}</b></p>
+     *
+     * @param queryName    A {@link String} describing the query to be executed
+     * @param query        The {@code query} to be sent
+     * @param responseType The {@link ResponseType} used for this query
+     * @param <R>          The response class contained in the given {@code responseType}
+     * @param <Q>          The query class
+     * @return A {@link Mono} containing the query result as dictated by the given {@code responseType}
+     */
+    <R, Q> Mono<R> query(String queryName, Q query, ResponseType<R> responseType);
+
+    /**
+     * Uses given Publisher of query messages to send incoming queries away. Queries will be sent sequentially - once a
+     * result of Nth query arrives, (N + 1)th query is dispatched.
+     *
+     * @param queries a Publisher stream of queries to be dispatched
+     * @return a Flux of query results. An ordering of query results corresponds to an ordering of queries being
+     * dispatched
+     *
+     * @see #query(String, Object, ResponseType)
+     * @see Flux#concatMap(Function)
+     */
+    default Flux<Object> query(Publisher<QueryMessage<?, ?>> queries) {
+        return Flux.from(queries)
+                   .concatMap(q -> query(q.getQueryName(), q.getPayload(), q.getResponseType()));
+    }
+
+    /**
+     * Sends the given {@code query} over the {@link QueryBus}, expecting a response in the form of {@code responseType}
+     * from several sources. The returned {@link Flux} is completed when a {@code timeout} occurs or when all possible
+     * results are received. The query name will be derived from the provided {@code query}. Execution may be
+     * asynchronous, depending on the {@code QueryBus} implementation.
+     * <p><b>Do note that the {@code query} will not be dispatched until there is a subscription to the resulting {@link
+     * Mono}</b></p>
+     * <b>Note</b>: Any {@code null} results will be filtered out by the {@link ReactorQueryGateway}. If you require
+     * the {@code null} to be returned, we suggest using {@code QueryBus} instead.
+     *
+     * @param query        The {@code query} to be sent
+     * @param responseType The {@link ResponseType} used for this query
+     * @param timeout      A timeout of {@code long} for the query
+     * @param timeUnit     The selected {@link TimeUnit} for the given {@code timeout}
+     * @param <R>          The response class contained in the given {@code responseType}
+     * @param <Q>          The query class
+     * @return A {@link Flux} containing the query results as dictated by the given {@code responseType}
+     */
+    default <R, Q> Flux<R> scatterGather(Q query, ResponseType<R> responseType, long timeout, TimeUnit timeUnit) {
```

Comment:
Is it possible to use [Duration](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html) instead of `long timeout, TimeUnit timeUnit` for the public interface and convert/split to `TimeUnit` and `long` for `QueryBus`'s needs?

Reply:
yes! good suggestion!